src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -709,77 +709,163 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
+                                biBitCount = bihReader.ReadUInt16(),
+                                biCompression = bihReader.ReadUInt32(),
+                                biSizeImage = bihReader.ReadUInt32(),
+                                biXPelsPerMeter = bihReader.ReadInt32(),
+                                biYPelsPerMeter = bihReader.ReadInt32(),
+                                biClrUsed = bihReader.ReadUInt32(),
+                                biClrImportant = bihReader.ReadUInt32()
+                            };
+
+                            IconImage iidata = new IconImage
+                            {
+                                iconHeader = bih
+                            };
 
-            iconDir.idType = reader.ReadUInt16();
-            if (iconDir.idType != 1) //must be 1
-                throw new System.ArgumentException("Invalid Argument", nameof(stream));
+                            // Read the number of colors used and corresponding memory occupied by
+                            // color table. Fill this memory chunk into rgbquad[]
+                            // number of colors = 2 ^ biBitCount
+                            int numColors = 0;
+                            if (bih.biBitCount < 24)
+                                numColors = (int)(1 << bih.biBitCount);
 
-            ushort dirEntryCount = reader.ReadUInt16();
-            imageData = new ImageData[dirEntryCount];
-            iconDir.idCount = dirEntryCount;
-            iconDir.idEntries = new IconDirEntry[dirEntryCount];
-            bool sizeObtained = false;
-            // now read in the IconDirEntry structures
-            for (int i = 0; i < dirEntryCount; i++)
-            {
-                IconDirEntry ide;
-                ide.width = reader.ReadByte();
-                ide.height = reader.ReadByte();
-                ide.colorCount = reader.ReadByte();
-                ide.reserved = reader.ReadByte();
-                ide.planes = reader.ReadUInt16();
-                ide.bitCount = reader.ReadUInt16();
-                ide.bytesInRes = reader.ReadUInt32();
-                ide.imageOffset = reader.ReadUInt32();
-
-                // Vista 256x256 icons points directly to a PNG bitmap
-                // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
-                // and we ignore them just like MS does (at least up to fx 2.0) 
-                // Added: storing data so it can be saved back
-                if ((ide.width == 0) && (ide.height == 0))
-                    ide.ignore = true;
-                else
-                    ide.ignore = false;
+                            iidata.iconColors = new uint[numColors];
+                            for (int j = 0; j < numColors; j++)
+                                iidata.iconColors[j] = bihReader.ReadUInt32();
 
-                iconDir.idEntries[i] = ide;
+                            try
+                            {
```

Comment:
Also added a try catch and checked value checks to catch an overflow exception as that is how this bug was happening.